proposals/generate_validating_admission_policies.md

```diff
@@ -0,0 +1,474 @@
+   + [Match/Exclude Resources](#matchexclude-resources)
+   + [Validate Rules](#validate-rules)
+   + [Audit Annotations](#audit-annotations)
+   + [Parameter Resources](#parameter-resources)
+   + [Limitations](#limitations-1)
+- [Next Steps](#next-steps)
+
+## Introduction
+Kyverno policies support CEL expressions for resources validation. The goal is to generate Kubernetes validating admission policies and validating admission binding from the information provided by Kyverno policies.
+
+- In case a new Kyverno policy is created, a validating admission policy and its binding are generated. 
+- In case Kyverno policy is updated, the corresponding VAP and its binding needs to be updated accordingly. 
+- In case Kyverno policy is deleted, the corresponding VAP and its binding are deleted.
+
+Once validating admission policy and its binding are generated, the API server will handle the resource validation. 
+
+An example of Kyverno policy from which validating admission policy and its binding will be generated:
+```yaml=
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: check-deployment-replicas
+spec:
+  validationFailureAction: Enforce
+  background: false
+  rules:
+    - name: deployment-replicas
+      match:
+        any:
+        - resources:
+            kinds:
+              - Deployment
+      validate:
+        cel:
+          expressions:
+            - expression: "object.spec.replicas < 5"
+              message:  "Deployment spec.replicas must be less than 5”
+```
+
+## Motivation
+Currently, Kyverno engine supports CEL expressions for resource validation. However, validating admission policy is an alternative solution to validating admission webhooks. Therefore, it's better to make Kyverno generate validating admission policies so that resource validation will be handled by the API server instead.
+
+Users can either let Kyverno engine to validate resources using CEL expressions or let the API server to validate resources and this is done by allowing Kyverno to generate validating admission policies and their bindings from a Kyverno policy containing a CEL subrule.
+
+## Proposal
+- A new controller `validatingadmissionpolicy-controller` will be created. It will watch the following:
+  1. Kyverno policies
+  2. Validating admission policies
+  3. Validating admission policy bindings
+
+- When a policy event happens, the policy key is enqueued. The controller will create/update/delete validating admission policies and their bindings accordingly.
+
+- When a validating admission policy or validating admission policy binding event happens, the policy key is deduced from owner references and it's enqueued. The controller will reset the changes done in either VAP or VAP binding.
+
+- A flag/toggle will be introduced in the controller so that user can either let Kyverno engine handles CEL expressions or let api server handles it by generating VAPs and their bindings.
```

Comment:
The default is on (generate VAPs and bindings)?

Reply:
The default is off because clusters with versions lower than 1.26 don't have validating admission policies so I make it off as default and if the user has the prerequisites to use VAPs then he can enable it.

The minimum requirements are:
1. Kubernetes >= v1.26
2. ValidatingAdmissionPolicy feature gate is enabled.
3. admissionregistration.k8s.io/v1alpha1 API is enabled.